plugins/destination/clickhouse/client/migrate.go

```diff
@@ -3,47 +3,84 @@ package client
 	"context"
 	"fmt"
+	"sort"
+	"strings"
 
 	"github.com/cloudquery/cloudquery/plugins/destination/clickhouse/queries"
 	"github.com/cloudquery/plugin-sdk/schema"
 	"github.com/cloudquery/plugin-sdk/specs"
-	"golang.org/x/sync/errgroup"
 )
 
 // Migrate relies on the CLI/client to lock before running migration.
 func (c *Client) Migrate(ctx context.Context, tables schema.Tables) error {
-	current, err := c.getTableDefinitions(ctx)
+	currentSchema, err := c.getTableDefinitions(ctx)
 	if err != nil {
 		return err
 	}
 
-	const maxConcurrentMigrate = 10
-	eg, ctx := errgroup.WithContext(ctx)
-	eg.SetLimit(maxConcurrentMigrate)
+	newSchema := queries.NormalizedTables(tables)
+	if c.spec.MigrateMode != specs.MigrateModeForced {
+		nonSafeMigratableTables, changes := c.nonAutoMigrableTables(newSchema, currentSchema)
+		if len(nonSafeMigratableTables) > 0 {
+			return fmt.Errorf("tables %s with changes %v require force migration. use 'migrate_mode: forced'", strings.Join(nonSafeMigratableTables, ","), changes)
+		}
+	}
 
-	for _, table := range queries.NormalizedTables(tables) {
+	for _, table := range newSchema {
 		table := table
-		eg.Go(func() (err error) {
-			c.logger.Info().Str("table", table.Name).Msg("Migrating table started")
-			defer func() {
-				c.logger.Err(err).Str("table", table.Name).Msg("Migrating table done")
-			}()
-
-			if len(table.Columns) == 0 {
-				c.logger.Warn().Str("table", table.Name).Msg("Table with no columns, skip")
-				return nil
-			}
+		c.logger.Info().Str("table", table.Name).Msg("Migrating table started")
+		if len(table.Columns) == 0 {
+			c.logger.Warn().Str("table", table.Name).Msg("Table with no columns, skip")
+			return nil
+		}
 
-			curr := current.Get(table.Name)
-			if curr == nil {
-				return c.createTable(ctx, table)
-			}
+		current := currentSchema.Get(table.Name)
+		if current == nil {
+			return c.createTable(ctx, table)
+		}
+
+		err := c.autoMigrate(ctx, table, current)
+		if err != nil {
+			return err
+		}
+		c.logger.Err(err).Str("table", table.Name).Msg("Migrating table done")
+	}
 
-			return c.autoMigrateTable(ctx, curr, table)
-		})
+	return nil
+}
+
+func (c *Client) nonAutoMigrableTables(tables schema.Tables, currentTables schema.Tables) ([]string, [][]schema.TableColumnChange) {
+	var result []string
+	var tableChanges [][]schema.TableColumnChange
+	for _, t := range tables {
+		current := currentTables.Get(t.Name)
+		if current == nil {
+			continue
+		}
+		changes := t.GetChanges(current)
+		if !c.canSafeMigrate(changes) {
+			result = append(result, t.Name)
+			tableChanges = append(tableChanges, changes)
+		}
 	}
+	return result, tableChanges
+}
+
+func (*Client) canSafeMigrate(changes []schema.TableColumnChange) bool {
+	for _, change := range changes {
+		if change.Type == schema.TableColumnChangeTypeAdd && change.Current.CreationOptions.NotNull {
+			return false
+		}
+
+		if change.Type == schema.TableColumnChangeTypeRemove && change.Previous.CreationOptions.NotNull {
```

Comment:
New/Current will be better that previous/current, as current should represent the current data schema, not the intended.
current = have in a have/want paradigm, basically.

Reply:
Those are SDK structs, unrelated to this PR. Do you mind opening an SDK issue to improve the API?